File: domain/src/main/java/com/ahn/domain/usecase/ExtractRepeatOperationUseCase.kt
package com.ahn.domain.usecase

import javax.inject.Inject

class ExtractRepeatOperationUseCase @Inject constructor() {
    private val operators = setOf('+', '-', '−', '×', '÷')

    /**
     * Extracts the last binary operator and its right-hand numeric operand from an arithmetic expression.
     *
     * @param expression The arithmetic expression to search.
     * @return A string containing the operator followed by the numeric operand (e.g., "+5.2") if a valid binary operator with a numeric right operand is found; `null` otherwise.
     */
    operator fun invoke(expression: String): String? {
        val operatorIndex = findLastBinaryOperatorIndex(expression)

        if (operatorIndex <= 0 || operatorIndex == expression.lastIndex) {
            return null
        }

        val operator = expression[operatorIndex].toString()
        val operand = expression.substring(operatorIndex + 1)

        if (operand.toDoubleOrNull() == null) {
            return null
        }

        return operator + operand
    }

    /**
     * Finds the index of the last binary operator in an arithmetic expression.
     *
     * An operator is treated as binary if it is one of the supported operator characters
     * and the character immediately before it is neither an operator nor `'('`.
     *
     * @param expression The arithmetic expression to search.
     * @return The index of the last binary operator, or `-1` if no binary operator is found.
     */
    private fun findLastBinaryOperatorIndex(expression: String): Int {
        return expression.indices.reversed().firstOrNull { index ->
            val current = expression[index]

            if (current !in operators) {
                return@firstOrNull false
            }

            if (index == 0) {
                return@firstOrNull false
            }

            val previous = expression[index - 1]

            previous !in operators && previous != '('
        } ?: -1
    }
}

File: ios/Packages/Domain/Sources/Domain/Calculator/UseCase/ExtractRepeatOperationUseCase.swift
import Foundation

public struct ExtractRepeatOperationUseCase: Sendable {
    private static let operators: Set<Character> = ["+", "-", "−", "×", "÷"]

    public init() {}

    public func callAsFunction(_ expression: String) -> String? {
        let chars = Array(expression)
        let operatorIndex = findLastBinaryOperatorIndex(chars)

        guard operatorIndex > 0, operatorIndex != chars.count - 1 else {
            return nil
        }

        let operatorSymbol = String(chars[operatorIndex])
        let operand = String(chars[(operatorIndex + 1)...])

        guard isNumericOperand(operand) else {
            return nil
        }

        return operatorSymbol + operand
    }

    private func isNumericOperand(_ operand: String) -> Bool {
        Double(operand.replacingOccurrences(of: "−", with: "-")) != nil
    }

    private func findLastBinaryOperatorIndex(_ chars: [Character]) -> Int {
        for index in stride(from: chars.count - 1, through: 0, by: -1) {
            let current = chars[index]

            guard Self.operators.contains(current) else { continue }
            guard index > 0 else { continue }

            let previous = chars[index - 1]

            if !Self.operators.contains(previous), previous != "(" {
                return index
            }
        }
        return -1
    }
}
